### PR TITLE
[SYCL] Aligned set_arg behaviour with SYCL specification

### PR DESCRIPTION
**Pull request by KarachunIvan**
According to SYCL 1.2.1 specification, rev. 7, paragraph 4.8.5:

template <typename T>
void set_arg(int argIndex, T &&arg)
...
The argument can be either a SYCL accessor, a SYCL sampler or
a trivially copyable and standard-layout C++ type.

Signed-off-by: Ivan Karachun <ivan.karachun@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/handler.hpp [hunk 1]
```diff
@@ -386,7 +386,11 @@ class __SYCL_EXPORT handler {
                        static_cast<int>(AccessTarget), ArgIndex);
   }
 
-  template <typename T> void setArgHelper(int ArgIndex, T &&Arg) {
+  template <typename T>
+  typename std::enable_if<std::is_trivially_copyable<T>::value &&
+                              std::is_standard_layout<T>::value,
```
Comment (bader) [inline sycl/include/CL/sycl/handler.hpp:391]:
```suggestion
                              std::is_standard_layout<T>::value,
```
I think this requirement was lifted in SYCL-2020. I suggest enforce it only for SYCL-1.2.1 version and keep `is_trivially_copyable` only otherwise.

Comment (KarachunIvan) [reply sycl/include/CL/sycl/handler.hpp]: Done

File: sycl/include/CL/sycl/handler.hpp [hunk 2]
```diff
@@ -386,7 +386,19 @@ class __SYCL_EXPORT handler {
                        static_cast<int>(AccessTarget), ArgIndex);
   }
 
-  template <typename T> void setArgHelper(int ArgIndex, T &&Arg) {
+  template <typename T> struct ShouldEnableSetArgHelper {
+    static constexpr bool value = std::is_trivially_copyable<T>::value
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/handler.hpp:390]: @bader  Should we also check for `std::is_trivially_destructible`?

Comment (bader) [reply sycl/include/CL/sycl/handler.hpp:390]: I don't think so, but I'm interested to see the case when it can matter.

Comment (romanovvlad) [reply sycl/include/CL/sycl/handler.hpp:390]: You are right. trivially copyable implies trivially destructible

File: sycl/include/CL/sycl/handler.hpp [hunk 3]
```diff
@@ -386,7 +386,19 @@ class __SYCL_EXPORT handler {
...
+#ifdef CL_SYCL_LANGUAGE_VERSION
+#if CL_SYCL_LANGUAGE_VERSION <= 121
```
Comment (mlychkov) [inline sycl/include/CL/sycl/handler.hpp:392]: May it be rewritten as:
`#if CL_SYCL_LANGUAGE_VERSION && CL_SYCL_LANGUAGE_VERSION <= 121`

Comment (KarachunIvan) [inline sycl/include/CL/sycl/handler.hpp:392]: BTW, I'm not sure about these ifdefs.
AFAIK, currently our compiler defines CL_SYCL_LANGUAGE_VERSION to be equal to 121 (since 1.2.1 is only supported).
https://github.com/intel/llvm/blob/4968e7c54d5030ee60b87efcfdfb21e6a7f0cfa6/clang/lib/Frontend/InitPreprocessor.cpp#L467
At the same time, SYCL 2020 defines new macro SYCL_LANGUAGE_VERSION. (5.6 Preprocessor directives and macros)

So, will the compiler define only SYCL_LANGUAGE_VERSION or it'll define two macros with the same value for SYCL-2020?
@bader, what do you think? Should I simplify these lines to this?
```suggestion
#if CL_SYCL_LANGUAGE_VERSION == 121
```

Comment (bader) [reply sycl/include/CL/sycl/handler.hpp:392]: This version looks okay to me.

Comment (KarachunIvan) [reply sycl/include/CL/sycl/handler.hpp]: Applied. Thanks for suggestion!

File: sycl/include/CL/sycl/handler.hpp [hunk 4]
```diff
@@ -796,13 +796,42 @@ class __SYCL_EXPORT handler {
...
+  template <typename U, typename T>
+  using is_same_type = std::is_same<remove_cv_ref_t<U>, remove_cv_ref_t<T>>;
+
+  template <typename T> struct ShouldEnableSetArg {
+    static constexpr bool value =
+        std::is_trivially_copyable<T>::value
+#if CL_SYCL_LANGUAGE_VERSION && CL_SYCL_LANGUAGE_VERSION <= 121
+            && std::is_standard_layout<T>::value
+#endif
+        || is_same_type<sampler, T>::value // Sampler
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/handler.hpp:812]: Why did you remove `remove_cv_ref_t<T>` ?

Comment (KarachunIvan) [reply sycl/include/CL/sycl/handler.hpp:812]: `remove_cv_ref_t` is used inside `is_same_type` for both types.
